Vendor a fork of async_executor to allow for deeper integration with bevy_tasks

james7132 opened the pull request:
# Objective
Improve the performance of Bevy's async task executor and clean up the code surrounding `TaskPool::scope`. Fixes #20421.

## Solution

 * Inline a copy of `async_executor`'s code
 * Move the TLS `LocalExecutor` and `ThreadExecutor` into the core `Executor` implementation, and poll their queues in `Executor::run`.
 * Use `!Send` types for local queues (`Mutex` -> `UnsafeCell`, `ConcurrentQueue` -> `VecDeque`).
 * Avoid extra `Arc` clones by using `&'static` references to thread-local queues.
 * Avoid extra contention on the global injector queue, and minimize extra allocations by opportunistically pushing onto thread-local queues when done on "executor-owned" threads.

This is a breaking change.

 * `ThreadExecutor` is now `ThreadSpawner` and forbids ticking and running the executor, only allowing spawning tasks to the target thread.
 * `Executor` and `LocalExecutor` wrappers are no longer public.

## Testing
I've only tested this against a few examples: 3d_scene and many_foxes. ~~This current implementation does seems to break animation systems.~~ EDIT: That seems to be #20383.

### Performance Testing
All of the benchmarks including a multithreaded schedule run are included below. Overall, this seems to be significant win whenever there are a very large number of tasks (i.e. the empty_archetypes/par_for_each` benchmarks), with some being upwards of 2x faster due to lower contention.

<details>

```
group                                                                                                     bevy_executor                            main
-----                                                                                                     -------------                            ----
added_archetypes/archetype_count/100                                                                      1.00     36.8±1.13µs        ? ?/sec      1.00     36.8±2.14µs        ? ?/sec
added_archetypes/archetype_count/1000                                                                     1.04   399.8±15.31µs        ? ?/sec      1.00   384.9±19.62µs        ? ?/sec
added_archetypes/archetype_count/10000                                                                    1.09     11.0±0.95ms        ? ?/sec      1.00     10.1±1.24ms        ? ?/sec
busy_systems/01x_entities_03_systems                                                                      1.00     18.6±0.56µs        ? ?/sec      2.16     40.3±1.18µs        ? ?/sec
busy_systems/01x_entities_09_systems                                                                      1.00     49.8±1.41µs        ? ?/sec      1.58     79.0±1.81µs        ? ?/sec
busy_systems/01x_entities_15_systems                                                                      1.00     84.0±1.46µs        ? ?/sec      1.39    116.4±2.48µs        ? ?/sec
busy_systems/03x_entities_03_systems                                                                      1.00     30.7±1.12µs        ? ?/sec      1.78     54.6±1.68µs        ? ?/sec
busy_systems/03x_entities_09_systems                                                                      1.00     83.3±2.12µs        ? ?/sec      1.50    125.3±3.44µs        ? ?/sec
busy_systems/03x_entities_15_systems                                                                      1.00    140.2±5.51µs        ? ?/sec      1.35    189.2±6.06µs        ? ?/sec
busy_systems/05x_entities_03_systems                                                                      1.00     50.8±1.00µs        ? ?/sec      1.53     77.7±6.65µs        ? ?/sec
busy_systems/05x_entities_09_systems                                                                      1.00    132.3±2.39µs        ? ?/sec      1.42    187.4±6.30µs        ? ?/sec
busy_systems/05x_entities_15_systems                                                                      1.00    217.3±5.15µs        ? ?/sec      1.29    279.7±5.10µs        ? ?/sec
contrived/01x_entities_03_systems                                                                         1.00      9.5±0.31µs        ? ?/sec      1.67     15.8±1.12µs        ? ?/sec
contrived/01x_entities_09_systems                                                                         1.00     23.1±0.65µs        ? ?/sec      1.56     36.1±5.34µs        ? ?/sec
contrived/01x_entities_15_systems                                                                         1.00     36.5±1.19µs        ? ?/sec      1.51     55.2±2.02µs        ? ?/sec
contrived/03x_entities_03_systems                                                                         1.00     21.1±1.03µs        ? ?/sec      1.70     35.9±2.03µs        ? ?/sec
contrived/03x_entities_09_systems                                                                         1.00     47.8±3.19µs        ? ?/sec      1.57     75.1±1.82µs        ? ?/sec
contrived/03x_entities_15_systems                                                                         1.00    77.7±13.55µs        ? ?/sec      1.54    119.7±9.98µs        ? ?/sec
contrived/05x_entities_03_systems                                                                         1.00     27.0±0.96µs        ? ?/sec      1.54     41.7±2.99µs        ? ?/sec
contrived/05x_entities_09_systems                                                                         1.00     67.1±4.48µs        ? ?/sec      1.57    105.7±5.74µs        ? ?/sec
contrived/05x_entities_15_systems                                                                         1.00   107.1±10.20µs        ? ?/sec      1.52   163.2±10.30µs        ? ?/sec
empty_archetypes/for_each/10                                                                              1.11  1792.3±56.15ns        ? ?/sec      1.00  1607.9±121.49ns        ? ?/sec
empty_archetypes/for_each/100                                                                             1.04  1752.9±71.42ns        ? ?/sec      1.00  1683.9±117.60ns        ? ?/sec
empty_archetypes/for_each/1000                                                                            1.00  1801.1±160.99ns        ? ?/sec     1.36      2.5±1.27µs        ? ?/sec
empty_archetypes/for_each/10000                                                                           1.00     13.3±0.55µs        ? ?/sec      1.40     18.6±1.52µs        ? ?/sec
empty_archetypes/iter/10                                                                                  1.02  1655.2±47.72ns        ? ?/sec      1.00  1624.9±138.81ns        ? ?/sec
empty_archetypes/iter/100                                                                                 1.00  1694.9±46.59ns        ? ?/sec      1.03  1746.4±202.97ns        ? ?/sec
empty_archetypes/iter/1000                                                                                1.00  1752.9±97.17ns        ? ?/sec      1.52      2.7±0.48µs        ? ?/sec
empty_archetypes/iter/10000                                                                               1.00      8.9±1.19µs        ? ?/sec      2.31     20.6±2.13µs        ? ?/sec
empty_archetypes/par_for_each/10                                                                          1.00      5.1±0.87µs        ? ?/sec      3.78     19.3±1.98µs        ? ?/sec
empty_archetypes/par_for_each/100                                                                         1.00      5.9±0.89µs        ? ?/sec      2.18     12.9±1.54µs        ? ?/sec
empty_archetypes/par_for_each/1000                                                                        1.00      5.3±0.39µs        ? ?/sec      3.68     19.5±1.76µs        ? ?/sec
empty_archetypes/par_for_each/10000                                                                       1.00     21.4±1.09µs        ? ?/sec      1.66     35.5±1.46µs        ? ?/sec
empty_commands/0_entities                                                                                 1.00      6.1±0.17ns        ? ?/sec      1.05      6.4±0.73ns        ? ?/sec
empty_systems/0_systems                                                                                   1.02     10.4±0.47ns        ? ?/sec      1.00     10.1±0.49ns        ? ?/sec
empty_systems/1000_systems                                                                                1.47   603.4±38.38µs        ? ?/sec      1.00   411.6±12.47µs        ? ?/sec
empty_systems/100_systems                                                                                 1.20     44.4±2.15µs        ? ?/sec      1.00     37.0±0.94µs        ? ?/sec
empty_systems/10_systems                                                                                  1.00      4.3±0.76µs        ? ?/sec      2.33     10.1±0.56µs        ? ?/sec
empty_systems/2_systems                                                                                   1.00  1964.4±158.36ns        ? ?/sec     2.94      5.8±0.96µs        ? ?/sec
empty_systems/4_systems                                                                                   1.00      2.6±0.13µs        ? ?/sec      3.46      8.9±0.92µs        ? ?/sec
for_each_par_iter/threads/1                                                                               1.00     11.1±0.17ms        ? ?/sec      1.89     20.9±1.50ms        ? ?/sec
for_each_par_iter/threads/16                                                                              1.01      3.1±0.10ms        ? ?/sec      1.00      3.1±0.10ms        ? ?/sec
for_each_par_iter/threads/2                                                                               1.00      7.6±0.14ms        ? ?/sec      1.49     11.4±0.29ms        ? ?/sec
for_each_par_iter/threads/32                                                                              1.01      2.2±0.04ms        ? ?/sec      1.00      2.2±0.06ms        ? ?/sec
for_each_par_iter/threads/4                                                                               1.00      4.8±0.16ms        ? ?/sec      1.61      7.6±0.16ms        ? ?/sec
for_each_par_iter/threads/8                                                                               1.00      4.6±0.74ms        ? ?/sec      1.03      4.7±0.10ms        ? ?/sec
many_maps_iter                                                                                            1.34     30.7±6.03ms        ? ?/sec      1.00     23.0±1.02ms        ? ?/sec
many_maps_par_iter/threads/1                                                                              1.00     11.6±0.37ms        ? ?/sec      1.83     21.2±1.42ms        ? ?/sec
many_maps_par_iter/threads/16                                                                             1.00      2.1±0.08ms        ? ?/sec      1.50      3.1±0.11ms        ? ?/sec
many_maps_par_iter/threads/2                                                                              1.00      7.7±0.11ms        ? ?/sec      1.48     11.4±0.30ms        ? ?/sec
many_maps_par_iter/threads/32                                                                             1.00      2.0±0.03ms        ? ?/sec      1.09      2.2±0.02ms        ? ?/sec
many_maps_par_iter/threads/4                                                                              1.00      4.7±0.08ms        ? ?/sec      1.66      7.8±0.12ms        ? ?/sec
many_maps_par_iter/threads/8                                                                              1.00      3.1±0.13ms        ? ?/sec      1.58      4.9±0.14ms        ? ?/sec
par_iter_simple/hybrid                                                                                    1.00     81.6±4.33µs        ? ?/sec      1.01     82.3±6.10µs        ? ?/sec
par_iter_simple/with_0_fragment                                                                           1.13     51.7±4.35µs        ? ?/sec      1.00     45.6±3.59µs        ? ?/sec
par_iter_simple/with_1000_fragment                                                                        1.04     67.8±3.77µs        ? ?/sec      1.00     65.1±5.16µs        ? ?/sec
par_iter_simple/with_100_fragment                                                                         1.09     53.2±3.21µs        ? ?/sec      1.00     48.8±3.43µs        ? ?/sec
par_iter_simple/with_10_fragment                                                                          1.08     51.4±4.00µs        ? ?/sec      1.00     47.5±5.10µs        ? ?/sec
param/combinator_system/8_dyn_params_system                                                               1.00  1429.6±60.37ns        ? ?/sec      1.16  1651.4±217.47ns        ? ?/sec
param/combinator_system/8_piped_systems                                                                   1.00  1382.3±62.79ns        ? ?/sec      1.18  1633.6±52.53ns        ? ?/sec
param/combinator_system/8_variant_param_set_system                                                        1.00  1375.6±49.81ns        ? ?/sec      1.31  1801.9±1117.63ns        ? ?/sec
run_condition/no/1000_systems                                                                             1.14     38.8±3.10µs        ? ?/sec      1.00     34.2±0.42µs        ? ?/sec
run_condition/no/100_systems                                                                              1.09      2.2±0.13µs        ? ?/sec      1.00      2.0±0.07µs        ? ?/sec
run_condition/no/10_systems                                                                               1.15   301.7±16.65ns        ? ?/sec      1.00   263.0±10.48ns        ? ?/sec
run_condition/yes/1000_systems                                                                            1.00   427.3±58.83µs        ? ?/sec      1.27   542.9±39.23µs        ? ?/sec
run_condition/yes/100_systems                                                                             1.00     37.9±1.33µs        ? ?/sec      1.22     46.2±3.51µs        ? ?/sec
run_condition/yes/10_systems                                                                              1.00      5.2±0.40µs        ? ?/sec      1.41      7.3±0.56µs        ? ?/sec
run_condition/yes_using_query/1000_systems                                                                1.00   413.4±22.99µs        ? ?/sec      1.38   570.8±77.20µs        ? ?/sec
run_condition/yes_using_query/100_systems                                                                 1.00     40.6±2.63µs        ? ?/sec      1.06     43.0±3.78µs        ? ?/sec
run_condition/yes_using_query/10_systems                                                                  1.00      6.0±1.24µs        ? ?/sec      1.25      7.4±0.52µs        ? ?/sec
run_condition/yes_using_resource/1000_systems                                                             1.07   423.0±29.47µs        ? ?/sec      1.00   394.9±29.42µs        ? ?/sec
run_condition/yes_using_resource/100_systems                                                              1.04     38.8±2.51µs        ? ?/sec      1.00     37.3±1.54µs        ? ?/sec
run_condition/yes_using_resource/10_systems                                                               1.00      5.4±0.51µs        ? ?/sec      1.95     10.5±0.75µs        ? ?/sec
run_empty_schedule/MultiThreaded                                                                          1.00      9.7±0.14ns        ? ?/sec      1.43     13.9±0.25ns        ? ?/sec
run_empty_schedule/Simple                                                                                 1.00     10.7±0.10ns        ? ?/sec      1.37     14.7±0.50ns        ? ?/sec
run_empty_schedule/SingleThreaded                                                                         1.00     12.5±0.22ns        ? ?/sec      1.34     16.8±1.34ns        ? ?/sec
schedule/base                                                                                             1.00     18.1±1.10µs        ? ?/sec      2.18     39.6±1.89µs        ? ?/sec
```
</details>

##  Future Work

 - [x] ~~Implement the [`StaticExecutor` optimization](https://github.com/smol-rs/async-executor/pull/112) for the static usages of the TaskPools. Would likely need a new implementation of #12990~~ Done in #20649.
 - [ ] Add support for [`rayon::spawn_broadcast`](https://docs.rs/rayon/latest/rayon/fn.spawn_broadcast.html) style APIs for more aggressively waking up threads than what the current implementation allows.
 - [ ] Replace `ThreadSpawner` with a more integrated `Executor::spawn_to_thread(ThreadId, Fut)` API instead, and use the ThreadIds tracked by `NonSend` resources to schedule those systems.
 - [ ] Support non-async Tasks (i.e. `FnOnce`) for lower-overhead execution, would avoid some extra atomics and state management.
 - [ ] Create a custom `block_on` implementation that ticks the local and thread-locked queues.

james7132 committed:
Inline async_executor

james7132 committed:
Merge LocalExecutor into the Executor implementation

james7132 committed:
Merge ThreadEecutors into the new forked async_executor

james7132 committed:
Move stealer queues to TLS storage. Avoid extra Arc allocations

james7132 committed:
Queue tasks directly back onto local queues to avoid global injector queue contention

james7132 committed:
Use UnsafeCell instead of RefCell

james7132 committed:
Optimize access of thread locked tasks

james7132 committed:
Clean up unnecessary unsafe use

james7132 committed:
Address potential unsoundness with ThreadSpawner

james7132 committed:
Fix some CI errors

james7132 committed:
Update docs

james7132 committed:
Format TOML files

james7132 committed:
Make note on ThreadLocal soundness hole

james7132 committed:
Allow clippy warning

james7132 committed:
Fix typos

james7132 committed:
Try to fix CI outside of miri

james7132 committed:
Remove depnedency on concurrent-queue

james7132 committed:
Revert "Clean up unnecessary unsafe use"

This reverts commit 8f1eaa71bcd28e79676e5930f29a2afe7daa6b79.

james7132 committed:
Fix single-threaded builds

james7132 committed:
Fix builds for no_std builds

james7132 committed:
Shut up Clippy

james7132 committed:
Fix bevy_ecs builds

james7132 committed:
Whoops

james7132 committed:
Arc is only used when no_std

james7132 committed:
Remove now unused mentions of LocalExecutor

james7132 committed:
Remove unused import

james7132 committed:
Use pin_project_lite for web builds instead of pin_project

james7132 committed:
Fix TOML formatting

james7132 committed:
Make CatchUnwind pin_project_lite friendly

james7132 committed:
Add missing AssertUnwindSafe

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Another attempt at fixing CI

NthTensor commented:
Read through this again. Didn't do a full review because it's still marked as a draft, but looks great. I think we'll probably want to merge this before <https://github.com/bevyengine/bevy/pull/20352>, it seems like my work will stack nicely on top of yours.

james7132 commented:
'd rather not block a soundness fix on such a large and potentially controversial change. This PR can wait until #20352 is merged.

NthTensor commented:
Fine by me.

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Properly handle thread destruction and recycling

james7132 commented:
Some of them, particularly those under a hundred microseconds and/or do more than just run empty systems, are more prone to having noise affect them with thread wake up times being upwards of 10-30us each on my machine.

That said, some of them do deserve some more scrutiny to see if those differences are replicable. The new polling function could easily add additional overhead when there's zero work to do.

james7132 committed:
Switch to a solution that doesn't require TLS access in thread destructors

james7132 committed:
Try to provide a blocking solution to TaskPool::scope in single threaded spaces

james7132 committed:
Merge branch 'main' into bevy_executor

NthTensor commented:
Currently the single threaded scope crates its own executor instead of spawning tasks on the thread local executor. This may sometimes prevent some futures from running while the scope completes. Is that acceptable?

It certainly is simpler than trying to spawn them on the larger thread local scope, and I've been thinking about adopting it in the other PR.

james7132 committed:
CI fixes

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Fix up the build, and hopefully CI

james7132 committed:
Shut up Clippy

james7132 committed:
Remove test println

james7132 committed:
Fix for portable atomics

james7132 committed:
Fix for web builds

hymm reviewed (commented):
This is not a full review. Mostly just did a quick skim of the code.

I'm generally onboard with adopting the async executor code, so we can customize it to our needs better. Adding the ability to queue local tasks in this pr shows how much it simplifies some of the above logic. But if we're doing this we should probably steal at some of their tests and be running miri on bevy_tasks.

I wouldn't be surprised if some of the regressions are due to us now checking the "local queue" more often now, when it was a ration one tick of LOCAL_EXECUTOR to 100 ticks of the TaskPool executor before.

james7132 commented on crates/bevy_tasks/src/task_pool.rs:
```diff
@@ -1,20 +1,16 @@
+use async_task::FallibleTask;
 use bevy_platform::sync::Arc;
-use concurrent_queue::ConcurrentQueue;
+use crossbeam_queue::SegQueue;
```
ConcurrentQueue is, internally, equivalent to an enum over all of crossbeam-queue's variants. This just avoids the extra `match` on its variants whenever you do an operation on them. Their implementations are nearly identical (sans the one-shot case).

For example, unlike the unbounded version of ConcurrentQueue still returns a Result on push, even though it can never fail, while SegQueue 

I mostly did this to dedupe dependencies with the existing unconditional dependency on crossbeam-queue required for edge_executor. ConcurrentQueue is present in the dep tree from different features and higher up in the dependency tree, but bevy_tasks, as of right now unconditionally requires crossbeam-queue.

I could revert this and substitute it back in. I *think* concurrent-queue is no_std (+alloc) compatible?

hymm replied:
I'm fine with the change.

NthTensor commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -140,9 +123,8 @@ impl TaskPool {
         // Any usages of the references passed into `Scope` must be accessed through
         // the transmuted reference for the rest of this function.
 
-        let executor = LocalExecutor::new();
```
I'm also curious about the effect of using a global executor here without prioritization. When there's just a single global queue, if the number of scoped tasks is small relative to the number of general tasks, our `yield_now` trick might take a while to chew through all the scheduled work and execute the tasks that are actually causing it to block. I think, Ideally, you'd want to poll the scoped tasks first and then the non-scoped tasks.

james7132 replied:
> Dropping a [Task](https://docs.rs/async-task/latest/async_task/struct.Task.html) cancels it, which means its future won’t be polled again. To drop the [Task](https://docs.rs/async-task/latest/async_task/struct.Task.html) handle without canceling it, use [detach()](https://docs.rs/async-task/latest/async_task/struct.Task.html#method.detach) instead. To cancel a task gracefully and wait until it is fully destroyed, use the [cancel()](https://docs.rs/async-task/latest/async_task/struct.Task.html#method.cancel) method.

We could just hold onto the `Task` instead of detatching them. If the panic propagates through, they'll all be dropped and thus cancel them all.

I'm also not 100% convinced that just using a new executor is sufficient. If a scoped task awaits a task from the global executor, the scope will block indefinitely.

---

Re: prioritizing Scope tasks, since the local queue is a `VecDeque`, we could just `push_back` instead of `push_front` on schedule. One of the first follow-up PRs I (almost) have ready to go also adds custom task priorities which alter the scheduling behavior of tasks.

james7132 replied:
> We could just hold onto the Task instead of detatching them. If the panic propagates through, they'll all be dropped and thus cancel them all.

I just reworked the single-threaded scope to do this.

james7132 committed:
Reduce async type genreation, code indirection, and handle single-threaded cancellation

james7132 committed:
Fix docs and move expect attribute

james7132 committed:
Complete the comment

james7132 committed:
Add async_executor's tests

james7132 committed:
Run Miri on bevy_tasks in CI

james7132 committed:
Make stealing a non-blocking operation

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Cache-pad the thread locals and disable multithreaded polling when the feature is not enabled

james7132 committed:
Fix Miri job to run them in sequence

james7132 committed:
Fix up README and Clippy

james7132 committed:
It's clippy

james7132 committed:
Address UB in spawn_scoped_local from OOMs

james7132 committed:
Move expect to outer block

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Fix lint

james7132 committed:
Go back to using ConcurrentQueue

james7132 committed:
Don't panic

james7132 committed:
Formatting

james7132 committed:
is_multiple_of lint

james7132 committed:
Merge branch 'main' into bevy_executor

james7132 committed:
Minimize cfg blocks

james7132 committed:
Try getting rid of the main thread local executor tick system

james7132 commented:
> But if we're doing this we should probably steal at some of their tests and be running miri on bevy_tasks.

I've gone ahead and added these in and added it to CI. Also did the following cleanups:

 * Avoided panicking wherever possible, replacing some asserts with debug_asserts.
 * Removed the main thread local executor tick system. It's no longer necessary.
 * Made stealing a strictly non-blocking operation.
 * Added some extra cache-padding to the local queues to avoid accidental false sharing.
 * Switched back to using ConcurrentQueue and elected to change edge_executor to use it instead.
 * Replaced the `pin-project` dependency with the proc-macro free `pin-project-lite` that was already in the dependency tree.

james7132 committed:
Use a proper struct for Sleepers

james7132 committed:
Update module docs

james7132 committed:
Remove unused code

james7132 committed:
Cleanup using utilities we already have

james7132 committed:
Avoid the extra Waker clone when initially scheduling tasks

james7132 committed:
Clippy, formatting, and panic documentation

NthTensor commented:
As I'm starting my review of this, I want to leave some broad observations.

+ All of the changes on the bevy side of things look great. This will simplify moving stuff off the main thread and in many cases are identical to changes I have in my prototype forte integration branch.
+ Changing single-threaded scopes to store a list of tasks seems like a pretty good idea. Might be worth splitting that off and merging it for 0.16, assuming there's still time.
+ The vendored version of `async_tasks` has significant differences from the current crate. I'm pleased to see it, but these will still require a fairly in-depth review.

I think I'll be focusing on the vendored code going forward. The other changes all look pretty reasonable.

NthTensor reviewed (approved):
This is a lot of work, and is a substantial improvement to both `bevy_tasks` and imo the original `async_executor` code. The safety comments all look correct to me, and I can't see any issues.

The original `async_executor` leaves something to be desired in terms of documentation. I'd prefer if documented all the types and functions in the vendored module. But I won't block on that.

NthTensor commented:
Uh, for some reason my review comments don't seem to be loading.

Not sure why that is. Anyone else seeing them?

hymm commented:
> Uh, for some reason my review comments don't seem to be loading.
> 
> Not sure why that is. Anyone else seeing them?

I only see one comment on the code.

NthTensor commented:
Looks like github ate them. I can still see them in codespaces, so I guess I'll have to copy them over? Kind of annoying.

Give me a sec.

NthTensor commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1296 @@
+    }
+}
+
+/// A task spawner for a specific thread. Must be created by calling [`TaskPool::current_thread_spawner`]
+/// from the target thread.
+///
+/// [`TaskPool::current_thread_spawner`]: crate::TaskPool::current_thread_spawner
+#[derive(Clone, Debug)]
+pub struct ThreadSpawner<'a> {
+    thread_id: ThreadId,
+    target_queue: &'static ConcurrentQueue<Runnable>,
+    state: Arc<State>,
+    _marker: PhantomData<&'a ()>,
+}
```
As a nit, I found this name slightly confusing. It sounds like it spawns threads, not "a spanwer local to the thread".

james7132 replied:
I went ahead and renamed it `LocalTaskSpawner` and `MainThreadTaskSpawner` instead.

NthTensor replied:
That's much clearer to me.

NthTensor commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1296 @@
+        //   `Waker`.
+        let (runnable, task) = unsafe {
+            Builder::new()
+                .propagate_panic(true)
```
For some reason I thought this was the default behavior, but I looked it up and I was wrong.

This is correct, and I'll have to adjust forte to match.

NthTensor commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1296 @@
+
+                crate::cfg::multi_threaded! {
+                    if {
+                        // Try the local queue.
+                        if let Ok(r) = self.local_state.stealable_queue.pop() {
+                            return Some(r);
+                        }
+
+                        // Try stealing from the global queue.
+                        if let Ok(r) = self.state.queue.pop() {
+                            steal(&self.state.queue, &self.local_state.stealable_queue);
+                            return Some(r);
+                        }
+
+                        // Try stealing from other runners.
+                        if let Ok(stealer_queues) = self.state.stealer_queues.try_read() {
+                            // Pick a random starting point in the iterator list and rotate the list.
+                            let n = stealer_queues.len();
+                            let start = _rng.usize(..n);
+                            let iter = stealer_queues
+                                .iter()
+                                .chain(stealer_queues.iter())
+                                .skip(start)
+                                .take(n);
+
+                            // Remove this runner's local queue.
+                            let iter =
+                                iter.filter(|local| !core::ptr::eq(**local, &self.local_state.stealable_queue));
+
+                            // Try stealing from each local queue in the list.
+                            for local in iter {
+                                steal(*local, &self.local_state.stealable_queue);
+                                if let Ok(r) = self.local_state.stealable_queue.pop() {
+                                    return Some(r);
+                                }
+                            }
```
Nit, because this is from the original implementation, but I somewhat disagree with the ordering here. I'm pretty sure rayon takes from the local queue, then from peers, then from the global queue, whereas this is local-global-peers. The rational for this is that it's better to finish your current work (stuff already queued on workers) before you start something new (by pulling from the injector queue).

Can we run some benchmarks with alternative orderings at some point?

NthTensor commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1296 @@
+                }
+
+                crate::cfg::multi_threaded! {
+                    if {
```
Since the else clause is empty, this if shouldn't be needed. Pretty sure you can just do `crate::cfg::multi_threaded! { ... }`.

james7132 committed:
Rename ThreadSpawner -> LocalTaskSpawner

james7132 committed:
Fix macro

NthTensor commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1296 @@
+}
+
+/// Steals some items from one queue into another.
+fn steal<T>(src: &ConcurrentQueue<T>, dest: &ConcurrentQueue<T>) {
+    // Half of `src`'s length rounded up.
+    let mut count = src.len();
+
+    if count > 0 {
+        if let Some(capacity) = dest.capacity() {
+            // Don't steal more than fits into the queue.
+            count = count.min(capacity- dest.len());
+        }
+
+        // Steal tasks.
+        for _ in 0..count {
+            let Ok(val) = src.pop() else { break };
+            assert!(dest.push(val).is_ok());
+        }
+    }
+}
```
Agreed.

hymm reviewed (commented):
Looks pretty good. I have some questions about some of the code I don't quite understand and still want to do some manual testing before I approve.

james7132 commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1294 @@
+// See: https://github.com/Amanieu/thread_local-rs/issues/75
+static THREAD_LOCAL_STATE: ThreadLocal<ThreadLocalState> = ThreadLocal::new();
+
+pub(crate) fn install_runtime_into_current_thread(executor: &Executor) {
```
The flag is not set and it's not considered an executor owned thread. It would fail the check inside schedule and default to the executor's global injector queue. This is primarily a way to avoid scheduling tasks onto the stealable queues of threads that are only momentarily calling `run` like the main or render threads.

james7132 commented on crates/bevy_tasks/src/bevy_executor.rs:
```diff
@@ -0,0 +1,1294 @@
+}
+
+struct LocalQueue {
+    local_queue: VecDeque<Runnable>,
```
As of right now, nothing. I don't see this as a particularly issue since that shouldn't be a soundness issue, and in the followup PRs, there may no longer be multiple executors to worry about.

hymm replied:
I was trying to come up with a scenario where this might cause problems since the lifetimes of the executors are different, but I couldn't come up with anything. It does make me nervous, but if multiple executors are going away this isn't worth worrying about.

james7132 committed:
Spin loop hints

james7132 commented on crates/bevy_tasks/src/task_pool.rs:
```diff
@@ -312,57 +298,39 @@ impl TaskPool {
     /// See [`Self::scope`] for more details in general about how scopes work.
     pub fn scope_with_executor<'env, F, T>(
         &self,
-        tick_task_pool_executor: bool,
```
This may very well be a regression there. The `run` call below would add the main/render thread to the list of executor threads without any block to what can run on them. Overall though, I don't think it's a great idea to be blocking those threads from running ComputeTaskPool tasks. Constrained systems with fewer cores (4 or fewer) would likely see underutilization of their hardware. Your suggestion of moving to async schedule runners to avoid the owning thread from hard-blocking might be a better approach. This may require "merging" adjacent schedules though.

> I think the ideal would be that the thread that the scope runs on would only run the tasks spawned on the scope, but not sure how feasible that would be. (Alternatively we could move to using async scope and systems to allow finishing the scope on any thread.)

With this PR, there's a softer prioritization. Tasks will be scheduled locally to the current thread, if possible, so assuming no other thread steals those tasks, the scope will be the one to execute them.

hymm replied:
If we merge this we should make an issue to look into fixing this as this will increase frame variance.

james7132 committed:
Merge branch 'main' into bevy_executor

hymm reviewed (approved):
First Schedule:
<img width="1053" height="267" alt="image" src="https://github.com/user-attachments/assets/2ca524f9-9074-4458-8641-1617ae931c37" />

This is the histogram for the First schedule. Yellow is this pr. The tail of yellow bad times is due to the main thread running render schedule systems.

Main schedule:
<img width="1057" height="265" alt="image" src="https://github.com/user-attachments/assets/01fa2e98-81f6-4141-9f63-9051718302ee" />

This problem extends into the main schedule. But we don't see this problem at the frame level since we're render bound.

Frame:
<img width="1069" height="319" alt="image" src="https://github.com/user-attachments/assets/8569ded2-949d-4db6-8ea4-ddbd5909e593" />

The overall execution time hasn't changed much, but the overall distribution is tighter.

We do see improvements in the schedule execution if we ignore the long tail problems, when the timing is broken down more. So I think this is worth merging it. I think if we made a async scope and a Schedule::run_async then the wins would be more clear. (might also need async systems to run schedules).  That work is blocked by removing nonsend resources.

Approving, but the hang on ci needs to be fixed.

james7132 commented on crates/bevy_tasks/src/edge_executor.rs:
```diff
@@ -97,6 +95,24 @@ impl<'a, const C: usize> Executor<'a, C> {
+        F: Future + 'a,
+        F::Output: 'a,
+    {
+        // SAFETY: Original implementation missing safety documentation
```
This was copied over from the other edge_executor spawn_* variants. It otherwise was missing the safety justification..

alice-i-cecile reviewed (approved):
I think this is the right direction: we need more control over this area of the code base if we're going to reduce system / task overhead, and even these initial improvements are quite nice.

I intend to merge this as we have a consensus of experts, but I'll [ping maintainers](https://discord.com/channels/691052431525675048/692572690833473578/1424821262940766415) and check the temperature first.
